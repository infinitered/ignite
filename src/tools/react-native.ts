import { GluegunToolbox } from "gluegun"
import { children } from "./filesystem-ext"
import { boolFlag } from "./flag"

export const isAndroidInstalled = (toolbox: GluegunToolbox): boolean => {
  const androidHome = process.env.ANDROID_HOME
  const hasAndroidEnv = !toolbox.strings.isBlank(androidHome)
  const hasAndroid = hasAndroidEnv && toolbox.filesystem.exists(`${androidHome}/tools`) === "dir"

  return Boolean(hasAndroid)
}

type CopyBoilerplateOptions = {
  boilerplatePath: string
  targetPath: string
  excluded: Array<string>
  overwrite?: boolean
}

/**
 * Copies the boilerplate over to the destination folder.
 *
 */
export async function copyBoilerplate(toolbox: GluegunToolbox, options: CopyBoilerplateOptions) {
  const { filesystem } = toolbox
  const { copyAsync, path } = filesystem

  // ensure the destination folder exists
  await filesystem.dirAsync(options.targetPath)

  // rather than copying everything wholesale, let's check what's in the boilerplate folder
  // and copy over everything except stuff like lockfiles and node_modules
  // just to make it faster, y'know? Don't want to copy unnecessary stuff
  const filesAndFolders = children(options.boilerplatePath, true)
  const copyTargets = filesAndFolders.filter(
    (file) => !options.excluded.find((exclusion) => file.includes(exclusion)),
  )

  const { overwrite } = options
  // kick off a bunch of copies
  const copyPromises = copyTargets.map((fileOrFolder) =>
    copyAsync(path(options.boilerplatePath, fileOrFolder), path(options.targetPath, fileOrFolder), {
      ...(overwrite && { overwrite }),
    }),
  )

  // copy them all at once
  return Promise.all(copyPromises)
}

export async function renameReactNativeApp(
  toolbox: GluegunToolbox,
  oldName: string,
  newName: string,
  oldBundleIdentifier: string,
  newBundleIdentifier: string,
) {
  const { parameters, filesystem, print, strings } = toolbox
  const { kebabCase } = strings
  const { path } = filesystem

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  // lower case stuff
  const oldnamelower = oldName.toLowerCase()
  const newnamelower = newName.toLowerCase()

  // kebab case
  const oldnamekebab = kebabCase(oldName)
  const newnamekebab = kebabCase(newName)

  // SCREAMING_SNAKE_CASE
  const oldnamesnake = oldnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()
  const newnamesnake = newnamelower.replace(/[^a-z0-9]/g, "_").toUpperCase()

  async function rename(oldFile: string, newFile: string) {
    log(`Renaming ${oldFile} to ${newFile}`)
    return filesystem.renameAsync(oldFile, newFile)
  }

  // rename files and folders

  // prettier-ignore
  await Promise.allSettled([
    rename(`ios/${oldName}.xcodeproj/xcshareddata/xcschemes/${oldName}.xcscheme`, `${newName}.xcscheme`),
    rename(`ios/${oldName}Tests/${oldName}Tests.m`, `${newName}Tests.m`),
    rename(`ios/${oldName}/${oldName}-Bridging-Header.h`, `${newName}-Bridging-Header.h`),
    rename(`ios/${oldName}/${oldName}.entitlements`, `${newName}.entitlements`),
    rename(`ios/${oldName}.xcworkspace`, `${newName}.xcworkspace`),
    rename(`ios/${oldName}`, `${newName}`),
  ])

  // these we delay to avoid race conditions
  await Promise.allSettled([
    rename(`ios/${oldName}Tests`, `${newName}Tests`),
    rename(`ios/${oldName}.xcodeproj`, `${newName}.xcodeproj`),
  ])

  // if the bundle identifier / android package name changed,
  // we need to move everything to the new folder structure
  const oldPath = oldBundleIdentifier.replace(/\./g, "/")
  const newPath = newBundleIdentifier.replace(/\./g, "/")

  if (oldBundleIdentifier !== newBundleIdentifier) {
    log(`Renaming bundle identifier to ${newBundleIdentifier}`)

    // move everything at the old bundle identifier path to the new one
    await Promise.allSettled([
      filesystem.moveAsync(
        `android/app/src/main/java/${oldPath}`,
        `android/app/src/main/java/${newPath}`,
      ),
      filesystem.moveAsync(
        `android/app/src/debug/java/${oldPath}`,
        `android/app/src/debug/java/${newPath}`,
      ),
      filesystem.moveAsync(
        `android/app/src/release/java/${oldPath}`,
        `android/app/src/release/java/${newPath}`,
      ),
    ])
  }

  // here's a list of all the files to patch the name in
  const filesToPatch = [
    `app.json`,
    `package.json`,
    `android/settings.gradle`,
    `android/app/_BUCK`,
    `android/app/BUCK`,
    `android/app/build.gradle`,
    `android/app/src/debug/java/${newPath}/ReactNativeFlipper.java`,
    `android/app/src/release/java/${newPath}/ReactNativeFlipper.java`,
    `android/app/src/main/AndroidManifest.xml`,
    `android/app/src/main/java/${newPath}/MainActivity.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/MainApplication.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/MainApplicationReactNativeHost.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/components/MainComponentsRegistry.java`,
    `android/app/src/main/java/${newPath}/newarchitecture/modules/MainApplicationTurboModuleManagerDelegate.java`,
    `android/app/src/main/jni/Android.mk`,
    `android/app/src/main/jni/MainApplicationTurboModuleManagerDelegate.h`,
    `android/app/src/main/jni/MainComponentsRegistry.h`,
    `android/app/src/main/res/values/strings.xml`,
    `ios/Podfile`,
    `ios/main.jsbundle`, // this file could just be regenerated, but this isn't bad to do
    `ios/${newName}/Info.plist`,
    `ios/${newName}.xcodeproj/project.pbxproj`,
    `ios/${newName}.xcodeproj/xcshareddata/xcschemes/${newName}.xcscheme`,
    `ios/${newName}.xcworkspace/contents.xcworkspacedata`,
    `ios/${newName}Tests/${newName}Tests.m`,
    `ios/${newName}/AppDelegate.mm`,
    `ios/${newName}/LaunchScreen.storyboard`,
  ]

  // patch the files
  await Promise.allSettled(
    filesToPatch.map(async (file) => {
      // no need to patch files that don't exist
      const exists = await filesystem.existsAsync(path(file))
      if (!exists) return

      const content = await filesystem.readAsync(path(process.cwd(), file), "utf8")

      log(`Patching ${file} - ${oldName} to ${newName} and variants`)

      // replace all instances of the old name and all its variants
      const newContent = content
        .replace(new RegExp(oldBundleIdentifier, "g"), newBundleIdentifier)
        .replace(new RegExp(oldnamekebab, "g"), newnamekebab)
        .replace(new RegExp(oldnamesnake, "g"), newnamesnake)
        .replace(new RegExp(oldName, "g"), newName)
        .replace(new RegExp(oldnamelower, "g"), newnamelower)

      // write the new content back to the file
      await filesystem.writeAsync(file, newContent, { atomic: true })
    }),
  )
}

export async function replaceMaestroBundleIds(
  toolbox: GluegunToolbox,
  oldBundleIdentifier: string,
  newBundleIdentifier: string,
) {
  const { parameters, filesystem, print } = toolbox
  const { path } = filesystem

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  // here's a list of all the maestro test files to fix the bundle id
  const TARGET_DIR = path(process.cwd())
  const filesToPatch = filesystem.cwd(TARGET_DIR).find({
    matching: `.maestro/**.yaml`,
    files: true,
    directories: false,
  })

  // patch the files
  await Promise.allSettled(
    filesToPatch.map(async (file) => {
      // no need to patch files that don't exist
      const exists = await filesystem.existsAsync(path(file))
      if (!exists) return

      const content = await filesystem.readAsync(path(process.cwd(), file), "utf8")

      log(`Patching ${file} - ${oldBundleIdentifier} to ${newBundleIdentifier} and variants`)

      // replace all instances of the placeholder bundle id with the new one
      const newContent = content.replace(new RegExp(oldBundleIdentifier, "g"), newBundleIdentifier)

      // write the new content back to the file
      await filesystem.writeAsync(file, newContent, { atomic: true })
    }),
  )
}

export function createExpoRouterScreenTemplate(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  try {
    const TARGET_DIR = filesystem.path(process.cwd())
    const filePath = filesystem.path(TARGET_DIR, "ignite/templates/screen/NAME.tsx.ejs")

    const EXPO_ROUTER_SCREEN_TPL = `import React, { FC } from "react"
import { observer } from "mobx-react-lite"
import { ViewStyle } from "react-native"
import { Screen, Text } from "src/components"

// @mst replace-next-line export default function <%= props.pascalCaseName %>Screen() {
export default observer(function <%= props.pascalCaseName %>Screen() {
  return (
    <Screen style={$root} preset="scroll">
      <Text text="<%= props.camelCaseName %>" />
    </Screen>
  )
// @mst replace-next-line }
})

const $root: ViewStyle = {
  flex: 1,
}
`
    filesystem.write(filePath, EXPO_ROUTER_SCREEN_TPL)
  } catch (e) {
    log(`Unable to write screen generator template.`)
  }
}

export function refactorExpoRouterReactotronCmds(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  try {
    const TARGET_DIR = filesystem.path(process.cwd())
    const reactotronConfigPath = filesystem.path(TARGET_DIR, "src/devtools/ReactotronConfig.ts")

    let reactotronConfig = filesystem.read(reactotronConfigPath)
    reactotronConfig = reactotronConfig
      .replace(/import { goBack, resetRoot, navigate }.*/g, 'import { router } from "expo-router"')
      .replace(/navigate\(route as any\).*/g, "router.push(route)")
      .replace(/goBack\(\).*/g, " router.back()")

    // this one gets removed entirely
    const customCommandToRemoveRegex =
      /reactotron\.onCustomCommand\({\s*title: "Reset Navigation State",\s*description: "Resets the navigation state",\s*command: "resetNavigation",\s*handler: \(\) => {\s*Reactotron\.log\("resetting navigation state"\)\s*resetRoot\({ index: 0, routes: \[\] }\)\s*},\s*}\),?\n?/g
    reactotronConfig = reactotronConfig.replace(customCommandToRemoveRegex, "")

    filesystem.write(reactotronConfigPath, reactotronConfig)
  } catch (e) {
    log(`Unable to update ReactotronConfig.`)
  }
}

export function updateExpoRouterSrcDir(toolbox: GluegunToolbox) {
  const { filesystem, parameters, print } = toolbox

  // debug?
  const debug = boolFlag(parameters.options.debug)
  const log = <T = unknown>(m: T): T => {
    debug && print.info(` ${m}`)
    return m
  }

  const TARGET_DIR = filesystem.path(process.cwd())
  const expoRouterFilesToFix = [
    "tsconfig.json",
    // has its own tsconfig, needs updating separately
    "test/i18n.test.ts",
    "test/setup.ts",
    "ignite/templates/model/NAME.ts.ejs",
    "ignite/templates/model/NAME.test.ts.ejs",
    "ignite/templates/component/NAME.tsx.ejs",
  ]
  expoRouterFilesToFix.forEach((file) => {
    const filePath = filesystem.path(TARGET_DIR, file)
    let fileContents = filesystem.read(filePath)
    try {
      fileContents = fileContents.replace(/app\//g, "src/")
      filesystem.write(filePath, fileContents)
    } catch (e) {
      log(`Unable to locate ${file}.`)
    }
  })
}

export function cleanupExpoRouterConversion(toolbox: GluegunToolbox, targetPath: string) {
  const { filesystem } = toolbox

  const workingDir = filesystem.cwd(targetPath)
  workingDir.cwd("src").remove("app.tsx")
  workingDir.move(
    workingDir.path("src", "screens", "ErrorScreen"),
    workingDir.path("src", "components", "ErrorBoundary"),
  )
  workingDir.remove("App.tsx")
  workingDir.remove(workingDir.path("ignite", "templates", "screen", "NAMEScreen.tsx.ejs"))
  workingDir.remove(workingDir.path("ignite", "templates", "navigator"))
  workingDir.remove(workingDir.path("src", "screens"))
  workingDir.remove(workingDir.path("src", "navigators"))
  workingDir.remove("app")
}
