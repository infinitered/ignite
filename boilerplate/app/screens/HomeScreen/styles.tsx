import { ImageStyle, TextStyle, ViewStyle } from 'react-native';
import { type ThemedStyle } from '@/theme';

export const $field: ThemedStyle<TextStyle> = () => ({
  borderRadius: 8,
  padding: 10,
  margin: 60,
});

export const $textInput: ThemedStyle<TextStyle> = () => ({
  margin: 20,
});

export const $view: ThemedStyle<ViewStyle> = ({ colors }) => ({
  padding: 20,
  backgroundColor: colors.background,
  borderRadius: 8,
  marginBottom: 80,
});

export const $listContentContainer: ThemedStyle<any> = ({ spacing }) => ({
  paddingHorizontal: spacing.lg,
});

export const $topContainer: ThemedStyle<ViewStyle> = ({ spacing }) => ({
  flexShrink: 1,
  flexGrow: 1,
  flexBasis: '57%',
  paddingHorizontal: spacing.lg,
});

export const $welcomeLogo: ThemedStyle<ImageStyle> = ({ spacing }) => ({
  height: 88,
  width: '100%',
  marginBottom: spacing.xxl,
});

export const $welcomeHeading: ThemedStyle<TextStyle> = ({ spacing }) => ({
  marginBottom: spacing.md,
});

export const $buttonAddItem: ThemedStyle<ViewStyle> = () => ({
  backgroundColor: 'white',
  borderRadius: 8,
  padding: 10,
  marginBottom: 50,
  alignItems: 'center',
  justifyContent: 'center',
});

export const $button: ThemedStyle<ViewStyle> = () => ({
  backgroundColor: 'white',
  borderRadius: 8,
  padding: 10,
  margin: 20,
  alignItems: 'center',
  justifyContent: 'center',
});
