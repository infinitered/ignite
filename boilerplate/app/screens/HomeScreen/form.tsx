import { FC } from 'react';
import { Text, Button, TextField } from '@/components';
import { type ThemedStyle } from '@/theme';
import { Item } from '@/models/Item';
import { translate } from '@/i18n';
import { View } from 'react-native';

import * as styles from './styles';

export const FormAddItem: FC<{
  addItem: (
    userId: string,
    item: { title: string; description: string },
  ) => void;
  themed: <ViewStyle>(styleOrStyleFn: ThemedStyle<ViewStyle>) => ViewStyle;
  userId: string;
  itemName: string;
  itemDescription: string;
  setItemName: (name: string) => void;
  setItemDescription: (description: string) => void;
  isFormValid: boolean;
  setIsFormValid: (isValid: boolean) => void;
  setDisplayForm: (display: boolean) => void;
}> = ({
  addItem,
  themed,
  userId,
  itemName,
  itemDescription,
  setItemName,
  setItemDescription,
  isFormValid,
  setIsFormValid,
  setDisplayForm,
}) => {
  return (
    <View style={themed(styles.$view)}>
      <Text
        testID="item-title"
        text={translate('homeScreen:itemName')}
        preset="heading"
        style={themed(styles.$textInput)}
      />
      <TextField
        testID="item-title-input"
        value={itemName}
        placeholder={translate('homeScreen:itemNamePlaceholder')}
        style={themed(styles.$field)}
        onChangeText={(text) => {
          setItemName(text);
          setIsFormValid(text.length > 1 && itemDescription.length > 1);
        }}
      />
      <Text
        testID="item-description"
        text={translate('homeScreen:itemDescription')}
        preset="heading"
        style={themed(styles.$textInput)}
      />
      <TextField
        testID="item-description-input"
        value={itemDescription}
        style={themed(styles.$field)}
        placeholder={translate('homeScreen:itemDescriptionPlaceholder')}
        onChangeText={(text) => {
          setItemDescription(text);
          setIsFormValid(itemName.length > 1 && text.length > 1);
        }}
      />
      <Button
        testID="ok-button"
        text={translate('common:ok')}
        onPress={() => {
          // Handle form submission
          setItemName('');
          setItemDescription('');
          setIsFormValid(false);
          setDisplayForm(false);
          addItem(userId, {
            title: itemName,
            description: itemDescription,
          });
        }}
        disabled={!isFormValid}
        style={themed(styles.$button)}
      />
      <Button
        testID="cancel-button"
        text={translate('common:cancel')}
        onPress={() => {
          // Handle form submission
          setItemName('');
          setItemDescription('');
          setIsFormValid(false);
          setDisplayForm(false);
        }}
        style={themed(styles.$button)}
      />
    </View>
  );
};

export const FormRemoveItem = ({
  themed,
  item,
  userId,
  removeItem,
  setDisplayRemoveForm,
  setItemSelectedForRemove,
}: {
  themed: <ViewStyle>(styleOrStyleFn: ThemedStyle<ViewStyle>) => ViewStyle;
  item: Item;
  userId: string;
  removeItem: (userId: string, item: Item) => void;
  setDisplayRemoveForm: (display: boolean) => void;
  setItemSelectedForRemove: (item: Item | null) => void;
}) => {
  return (
    <View style={themed(styles.$view)}>
      <Text
        testID="remove-item-description"
        text={translate('homeScreen:removeItem')}
        preset="heading"
      />
      <Button
        testID="yes-button"
        text={translate('common:yes')}
        onPress={() => {
          removeItem(userId, item);
          setDisplayRemoveForm(false);
          setItemSelectedForRemove(null);
        }}
        style={themed(styles.$button)}
      />
      <Button
        testID="no-button"
        text={translate('common:no')}
        onPress={() => {
          setDisplayRemoveForm(false);
          setItemSelectedForRemove(null);
        }}
        style={themed(styles.$button)}
      />
    </View>
  );
};
