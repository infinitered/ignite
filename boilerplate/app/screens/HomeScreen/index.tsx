import { observer } from 'mobx-react-lite';
import { FC, useEffect, useState, useCallback } from 'react';
import Modal from 'react-native-modal';
import { FlatList, Image, View } from 'react-native';

import { Text, Screen, Button } from '@/components';
import { $styles } from '@/theme';
import { useAppTheme } from '@/utils/useAppTheme';
import { Item } from '@/models/Item';
import i18n from 'i18next';

import { useHeader } from '../../utils/useHeader';
import { NativeListItem, ListItem } from '../List';
import { useStores } from '../../models';
import { AppStackScreenProps } from '../../navigators';
import { FormAddItem, FormRemoveItem } from './form';
import { translate } from '@/i18n';

import * as styles from './styles';

const welcomeLogo = require('../../../assets/images/spacebox.png');

interface HomeScreenProps extends AppStackScreenProps<'Home'> {}

export const HomeScreen: FC<HomeScreenProps> = observer(
  function HomeScreen(_props) {
    const { themed } = useAppTheme();
    const [displayForm, setDisplayForm] = useState(false);
    const [displayRemoveForm, setDisplayRemoveForm] = useState(false);
    const [itemSelectedForRemove, setItemSelectedForRemove] =
      useState<Item | null>(null);
    const [itemName, setItemName] = useState('');
    const [itemDescription, setItemDescription] = useState('');
    const [isFormValid, setIsFormValid] = useState(false);
    const [initItems, setInitItems] = useState(false);

    const [, updateState] = useState({});
    const forceUpdate = useCallback(() => updateState({}), []);

    const {
      authenticationStore: { logout, authFirstName, authLastName, userId },
      storageStore: {
        items,
        addItemAsync: addItem,
        removeItemAsync: removeItem,
        fetchItems,
      },
    } = useStores();

    useHeader(
      {
        rightTx: 'common:logOut',
        leftTx: 'common:lang',
        onLeftPress: () => {
          if (i18n.language.includes('en')) {
            i18n.changeLanguage('zh');
          } else {
            i18n.changeLanguage('en');
          }
          forceUpdate();
        },
        onRightPress: logout,
      },
      [logout, i18n.language],
    );

    useEffect(() => {
      const fetchData = async () => {
        if (!userId) {
          return;
        }
        await fetchItems(userId);
        setInitItems(true);
      };
      if (!initItems) {
        fetchData();
      }
    }, [fetchItems, userId, initItems]);

    return (
      <Screen preset="fixed" contentContainerStyle={$styles.flex1}>
        <View style={themed(styles.$topContainer)}>
          <Image
            style={themed(styles.$welcomeLogo)}
            source={welcomeLogo}
            resizeMode="contain"
          />
          <Text
            testID="welcome-name"
            style={themed(styles.$welcomeHeading)}
            text={`${translate('common:hello')} ${authFirstName} ${authLastName}`}
            preset="heading"
          />
          <FlatList<ListItem['item']>
            contentContainerStyle={themed(styles.$listContentContainer)}
            data={items}
            renderItem={({ item }) => {
              return (
                <NativeListItem
                  {...{
                    item,
                    setDisplayRemoveForm,
                    setItemSelectedForRemove,
                  }}
                />
              );
            }}
          />
          <Button
            testID="add-item-button"
            text={translate('homeScreen:addItem')}
            onPress={() => {
              setDisplayForm(!displayForm);
            }}
            style={themed(styles.$buttonAddItem)}
          />
          {displayForm && userId && (
            <Modal
              isVisible={displayForm}
              onBackdropPress={() => setDisplayForm(false)}
            >
              <FormAddItem
                userId={userId}
                addItem={addItem}
                themed={themed}
                itemName={itemName}
                itemDescription={itemDescription}
                setItemName={setItemName}
                setItemDescription={setItemDescription}
                isFormValid={isFormValid}
                setIsFormValid={setIsFormValid}
                setDisplayForm={setDisplayForm}
              />
            </Modal>
          )}
          {displayRemoveForm && itemSelectedForRemove && userId && (
            <Modal
              isVisible={displayRemoveForm}
              onBackdropPress={() => setDisplayRemoveForm(false)}
            >
              <FormRemoveItem
                themed={themed}
                userId={userId}
                item={itemSelectedForRemove}
                setDisplayRemoveForm={setDisplayRemoveForm}
                setItemSelectedForRemove={setItemSelectedForRemove}
                removeItem={removeItem}
              />
            </Modal>
          )}
        </View>
      </Screen>
    );
  },
);
