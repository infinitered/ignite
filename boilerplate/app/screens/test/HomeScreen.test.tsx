import { screen, render, fireEvent, act } from '@testing-library/react-native';

import { HomeScreen } from '../HomeScreen';
// Mock the necessary modules and functions
jest.mock('../../models', () => ({
  useStores: jest.fn(() => ({
    authenticationStore: {
      userId: '1',
    },
    storageStore: {
      fetchItems: () => Promise.resolve(),
    },
  })),
}));

jest.mock('@/utils/useAppTheme', () => ({
  useAppTheme: jest.fn(() => ({
    themed: () => (Component: any) => Component,
    theme: {
      colors: {},
    },
  })),
}));

jest.mock('../../utils/useHeader', () => ({
  useHeader: jest.fn(),
}));

jest.mock('i18next', () => ({
  changeLanguage: jest.fn(),
  language: 'en',
}));

describe('HomeScreen', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders the Home screen - Snapshot', () => {
    render(<HomeScreen navigation={{} as any} route={{} as any} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });

  it('should open the modal to add item and click on cancel button', async () => {
    const { getByTestId } = render(
      <HomeScreen navigation={{} as any} route={{} as any} />,
    );

    await act(async () => {
      // wait for the modal to be visible
      await new Promise((resolve) => setTimeout(resolve, 1000));
    });

    // Click on the add item button
    const addItemButton = getByTestId('add-item-button');
    fireEvent.press(addItemButton);

    // Add item title
    const itemTitleInput = getByTestId('item-title-input');
    fireEvent.changeText(itemTitleInput, 'Item 1');

    // Add item description
    const itemDescriptionInput = getByTestId('item-description-input');
    fireEvent.changeText(itemDescriptionInput, 'Item 1 description');

    // Click on the cancel button
    const cancelButton = getByTestId('cancel-button');
    fireEvent.press(cancelButton);
  });
});
