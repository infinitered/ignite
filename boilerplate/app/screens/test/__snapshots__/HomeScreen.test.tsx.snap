// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen renders the Home screen - Snapshot 1`] = `
<View
  style={
    [
      {
        "flex": 1,
        "height": "100%",
        "width": "100%",
      },
      {
        "backgroundColor": undefined,
      },
      {},
    ]
  }
>
  <View
    onLayout={[Function]}
    style={
      [
        [
          {
            "flex": 1,
          },
          undefined,
        ],
        {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "flex": 1,
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "stretch",
              "justifyContent": "flex-start",
            },
            {
              "justifyContent": "flex-end",
            },
            {
              "flex": 1,
            },
          ]
        }
      >
        <View
          style={[Function]}
        >
          <RCTImageView
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            internal_analyticTag={null}
            resizeMode="contain"
            source={
              [
                {
                  "testUri": "../../../assets/images/spacebox.png",
                },
              ]
            }
            style={
              [
                {
                  "height": undefined,
                  "width": undefined,
                },
                {
                  "overflow": "hidden",
                },
                [Function],
              ]
            }
          />
          <Text
            preset="heading"
            style={
              [
                {},
                [Function],
                undefined,
                undefined,
                [Function],
              ]
            }
            testID="welcome-name"
          >
            common:hello undefined undefined
          </Text>
          <RCTScrollView
            contentContainerStyle={[Function]}
            getItem={[Function]}
            getItemCount={[Function]}
            keyExtractor={[Function]}
            onContentSizeChange={[Function]}
            onLayout={[Function]}
            onMomentumScrollBegin={[Function]}
            onMomentumScrollEnd={[Function]}
            onScroll={[Function]}
            onScrollBeginDrag={[Function]}
            onScrollEndDrag={[Function]}
            removeClippedSubviews={false}
            renderItem={[Function]}
            scrollEventThrottle={0.0001}
            stickyHeaderIndices={[]}
            viewabilityConfigCallbackPairs={[]}
          >
            <View />
          </RCTScrollView>
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": false,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                [Function],
                [Function],
                false,
                false,
              ]
            }
            testID="add-item-button"
          >
            <Text
              style={
                [
                  {},
                  [Function],
                  undefined,
                  undefined,
                  [
                    [Function],
                    undefined,
                    false,
                    false,
                  ],
                ]
              }
            >
              homeScreen:addItem
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
